src/Microsoft.AspNet.OData.Shared/DeltaSetOfT.cs

```diff
@@ -253,6 +289,172 @@ internal DeltaSet<TStructuralType> CopyChangedValues(IODataAPIHandler apiHandler
+
+                try
+                {
+                    Dictionary<string, object> keyValues = new Dictionary<string, object>();
+                    keyValues = changedObj.ODataPath.GetKeys();
+
+                    bool containsKeyValue = false;
+
+                    foreach (string key in _keys)
+                    {
+                        keyValues.TryGetValue(key, out object value);
+
+                        if (value != null && !string.IsNullOrEmpty(value.ToString()))
+                        {
+                            containsKeyValue = true;
+                            continue;
+                        }
+                    }
+
+                    ODataAPIResponseStatus odataAPIResponseStatus;
+
+                    if (containsKeyValue)
+                    {
+                        odataAPIResponseStatus = await handler.TryGetAsync(keyValues, out original, out getErrorMessage);
+
+                        if (odataAPIResponseStatus != ODataAPIResponseStatus.Success && apiHandlerFactory != null)
+                        {
+                            ODataAPIHandler<TStructuralType> odataPathApiHandler = await apiHandlerFactory.GetHandlerAsync(changedObj.ODataPath) as ODataAPIHandler<TStructuralType>;
+
+                            if (odataPathApiHandler != null)
+                            {
+                                handler = odataPathApiHandler;
+                                odataAPIResponseStatus = await handler.TryGetAsync(keyValues, out original, out getErrorMessage);
+                            }
+                        }
+                    }
+                    else
+                    {
+                        odataAPIResponseStatus = ODataAPIResponseStatus.NotFound;
+                    }
+
+                    DeltaDeletedEntityObject<TStructuralType> deletedObj = changedObj as DeltaDeletedEntityObject<TStructuralType>;
+
+                    if (odataAPIResponseStatus == ODataAPIResponseStatus.Failure)
+                    {
+                        IDeltaSetItem deltaSetItem = changedObj;
+                        DataModificationExceptionType dataModificationExceptionType = new DataModificationExceptionType(operation);
+                        dataModificationExceptionType.MessageType = new MessageType { Message = getErrorMessage };
+
+                        deltaSetItem.TransientInstanceAnnotationContainer.AddResourceAnnotation(SRResources.DataModificationException, dataModificationExceptionType);
+                        deltaSet.Add(deltaSetItem);
+
+                        continue;
+                    }
+
+                    if (deletedObj != null)
+                    {
+                        operation = DataModificationOperationKind.Delete;
+
+                        if (odataAPIResponseStatus == ODataAPIResponseStatus.NotFound)
+                        {
+                            // Handle Failed Operation - Delete when the object doesn't exist.
```

Comment:
We are not deleting but we create a delete compensating action. Since this was a delete operation and we did not find what we were deleting then it means that it does not exist in the server and it needs to be deleted locally if it already exists.